haskell/src/Monocle/Api/Server.hs

```diff
@@ -29,14 +30,101 @@ configHealth = const $ pure response
   where
     response = ConfigPB.HealthResponse "api running"
 
+-- Mock to bind to function from another PR
+indexExist name = True
+
+crawlerExist indexName crawlerName = True
+
+apiKeyMatch indexName crawlerName apiKey = True
+
 crawlerAddDoc :: CrawlerPB.AddDocRequest -> AppM CrawlerPB.AddDocResponse
-crawlerAddDoc = undefined
+crawlerAddDoc request = do
+  Env {bhEnv = bhEnv} <- ask
+  let (CrawlerPB.AddDocRequest indexName crawlerName apiKey entity changes events) = request
+  case (indexExist indexName, crawlerExist indexName crawlerName, apiKeyMatch indexName crawlerName apiKey) of
+    (True, True, True) -> do
+      case toEntity entity of
```

Comment:
It seems like these checks can be rewritten using this trick: https://www.haskellforall.com/2021/05/the-trick-to-avoid-deeply-nested-error.html

In this case, that would look like something like this:

```haskell
orStop :: Maybe a -> b -> Either b a
orStop = undefined

getResp :: IO CrawlerPB.AddDocResponse
getResp = either (pure . toErrorResponse) go validateRequest
  where
    validateRequest :: Either CrawlerPB.AddDocError ()
    validateRequest = do
      index <- lookupIndex inderName `orStop` AddDocErrorAddUnknownIndex
      crawler <- lookupCrawler index crawlerName `orStop` AddDocErrorAddUnknownCrawler
      when (crawlerKey /= apiKey) (Left AddDocErrorAddUnknownApiKey)
      pure ()
    -- the validation doesn't produce anything, so we `_` the result here
    go _ = do
      I.indexChanges ...
```

Reply:
Here is an example for the search query validation: https://github.com/change-metrics/monocle/pull/426/commits/a952c4f925794610e1eaa8e32df05d5c01c34d73 . Notice that when all the validation happens in the Either, then we don't need to combine with `orStop`.

Though that is really cosmetic, your change LGTM, although the mock need to be removed :-)

Reply:
Thanks for the Either trick, It not strength forward for me as of now. I've tried but it's taking more time time than expected and I figured out I need to tackle function like lookupCrawler first.